### PR TITLE
Publish pnlppgi

### PR DESCRIPTION
**Pull request by lwyszomi**
@fyniac 
This is my proposition to publish custom reports, could you assign someone who can check this?

### CODE REVIEW

File: corehq/apps/users/models.py [hunk 1]
```diff
@@ -2376,3 +2376,59 @@ def get(self, user_id):
...
+        return False
+
+    def can_edit_commcare_users(self):
+        return False
+
+    def can_edit_locations(self):
+        return False
+
+    def can_edit_web_users(self):
+        return False
```
Comment (dimagimon) [inline corehq/apps/users/models.py]: W292 no newline at end of file

File: corehq/toggles.py [hunk 2]
```diff
@@ -1058,3 +1058,10 @@ def _commtrackify(domain_name, toggle_is_enabled):
     TAG_EXPERIMENTAL,
     [NAMESPACE_DOMAIN, NAMESPACE_USER]
 )
+
+PUBLISH_CUSTOM_REPORTS = StaticToggle(
+    'publish_custom_reports',
+    "Publish custom reports (No needed Authorization)",
+    TAG_EXPERIMENTAL,
+    [NAMESPACE_DOMAIN]
+)
```
Comment (dimagimon) [inline corehq/toggles.py]: W292 no newline at end of file

**Comment by lwyszomi** [comment]
@snopoke Hey, I talk with Lucina and she told me that you checked these changes. Also she told me that I should add separate page and remove export/email buttons. Could you give me more details about separate page (you want only new template?).

**Comment by lwyszomi** [comment]
@snopoke bump

File: corehq/apps/users/middleware.py [hunk 3]
```diff
@@ -41,4 +39,7 @@ def process_view(self, request, view_func, view_args, view_kwargs):
                     request.couch_user = InvalidUser()
                 if request.couch_user:
                     request.couch_user.current_domain = domain
+        elif 'domain' in view_kwargs and '/reports/custom/' in request.path\
```
Comment (snopoke) [inline corehq/apps/users/middleware.py]: I think this should be under `if request.user.is_anonymous() and 'domain' in view_kwargs:`

Comment (snopoke) [inline corehq/apps/users/middleware.py]: The path check should be more restrictive:
```python
request.path.startswith(u'/a/{}/reports/custom'.format(domain))
```

File: corehq/apps/reports/dispatcher.py [hunk 4]
```diff
@@ -269,6 +270,8 @@ def permissions_check(self, report, request, domain=None, is_navigation_check=Fa
             return False
         if not request.couch_user.is_active:
             return False
+        if isinstance(request.couch_user, PublicCouchUser) and self.prefix != 'custom_project_report':
+            return False
```
Comment (snopoke) [inline corehq/apps/reports/dispatcher.py]: this should be in the `CustomProjectReportDispatcher`

**Review by snopoke — COMMENTED** [review]
@benrudolph could you give this a once over - I feel like you have a better sense of this kind of thing after your work on anonymous web apps.

**Comment by snopoke** [comment]
@benrudolph the idea here is to make custom reports publically accessible.

@lwyszomi it would be good to be able to restrict this to only certain reports. I'm also not sure about the `PublicCouchUser` class - seems risky to have all permissions.

**Comment by benrudolph** [comment]
ah very interesting. will review

File: corehq/apps/reports/dispatcher.py (hunk 4 above)
Comment (benrudolph) [inline corehq/apps/reports/dispatcher.py]: +1

File: corehq/apps/users/middleware.py (hunk 3 above)
Comment (benrudolph) [inline corehq/apps/users/middleware.py]: +1

**Review by benrudolph — COMMENTED** [review]
Instead of making PublicCouchUser a couch doc that gets saved to the database, can we just make this a plain object? like AnonymousCouchUser (similar to Django's AnonymousUser). And then have it have a method that is `is_anonymous` that will return `True`. Then it's on the report's implementation to allow AnonymousCouchUser's to use the report. I would rather not give the anonymous user all permissions. That feels like that could accidentally go wrong.

**Comment by lwyszomi** [comment]
@benrudolph @snopoke Thanks for your feedback, I will implement all your suggestions.

File: corehq/apps/users/middleware.py [hunk 5]
```diff
@@ -1,16 +1,20 @@
...
+from corehq.apps.users.models import CouchUser, InvalidUser, AnonymousCouchUser
+from corehq.toggles import ANONYMOUS_WEB_APPS_USAGE, PUBLISH_CUSTOM_REPORTS
 
+SESSION_USER_KEY_PREFIX = "session_user_doc_%s"
 
 
-SESSION_USER_KEY_PREFIX = "session_user_doc_%s"
+def is_public_reports(view_kwargs, request):
+    return request.user.is_anonymous() and 'domain' in view_kwargs and \
+           request.path.startswith(u'/a/{}/reports/custom'.format(view_kwargs['domain'])) and \
```
Comment (dimagimon) [inline corehq/apps/users/middleware.py]: E126 continuation line over-indented for hanging indent

File: corehq/apps/users/middleware.py [hunk 6]
```diff
@@ -1,16 +1,20 @@
...
+from corehq.apps.users.models import CouchUser, InvalidUser, AnonymousCouchUser
+from corehq.toggles import ANONYMOUS_WEB_APPS_USAGE, PUBLISH_CUSTOM_REPORTS
 
+SESSION_USER_KEY_PREFIX = "session_user_doc_%s"
 
 
-SESSION_USER_KEY_PREFIX = "session_user_doc_%s"
+def is_public_reports(view_kwargs, request):
+    return request.user.is_anonymous() and 'domain' in view_kwargs and \
+       request.path.startswith(u'/a/{}/reports/custom'.format(view_kwargs['domain'])) and \
```
Comment (dimagimon) [inline corehq/apps/users/middleware.py]: E121 continuation line under-indented for hanging indent

**Comment by lwyszomi** [comment]
@benrudolph @snopoke I updated the code. I'm still working to public specific custom report.

File: corehq/apps/domain/decorators.py [hunk 7]
```diff
@@ -116,6 +116,8 @@ def _inner(req, domain, *args, **kwargs):
                     return DomainRequestView.as_view()(req, *args, **kwargs)
                 else:
                     raise Http404
+            elif 'reports/custom/' in req.path and PUBLISH_CUSTOM_REPORTS.enabled(domain_name):
```
Comment (snopoke) [inline corehq/apps/domain/decorators.py]: This should still be more restrictive (see my last comment about it)

File: corehq/apps/users/models.py [hunk 8]
```diff
@@ -2376,3 +2376,63 @@ def get(self, user_id):
...
+        return True
+
+    def is_domain_admin(self):
+        return False
+
+    def is_member_of(self, domain):
+        return True
+
+    def has_permission(self, domain, perm=None, data=None):
+        return True
```
Comment (snopoke) [inline corehq/apps/users/models.py]: Why does this need to be true?

**Comment by lwyszomi** [comment]
@snopoke @benrudolph I updated the PR and I added possibility to publish specific custom report.

**Comment by fyniac** [comment]
@benrudolph can you please look at this?

File: corehq/apps/locations/middleware.py [hunk 9]
```diff
@@ -30,7 +30,7 @@ def process_view(self, request, view_fn, view_args, view_kwargs):
         if not domain or not user or not user.is_member_of(domain):
             # This is probably some non-domain page or a test, let normal auth handle it
             request.can_access_all_locations = True
-        elif user.has_permission(domain, 'access_all_locations'):
+        elif user.has_permission(domain, 'access_all_locations') or user.is_anonymous:
```
Comment (benrudolph) [inline corehq/apps/locations/middleware.py]: might be safer to do `or (user.is_anonymous and PUBLISH_CUSTOM_REPORTS.enabled(domain)`

**Review by benrudolph — COMMENTED** [review]
ok one minor comment about a toggle check, and then i think g2g. sorry just want to be sure we're not accidentally opening up pages on HQ (don't think we are), but important to guard middleware

File: corehq/apps/locations/middleware.py [hunk 10]
```diff
@@ -30,7 +31,10 @@ def process_view(self, request, view_fn, view_args, view_kwargs):
         if not domain or not user or not user.is_member_of(domain):
             # This is probably some non-domain page or a test, let normal auth handle it
             request.can_access_all_locations = True
-        elif user.has_permission(domain, 'access_all_locations'):
+        elif user.has_permission(domain, 'access_all_locations') or (
+                    user.is_anonymous and
```
Comment (dimagimon) [inline corehq/apps/locations/middleware.py]: E126 continuation line over-indented for hanging indent

**Comment by lwyszomi** [comment]
@benrudolph updated

**Comment by benrudolph** [comment]
thanks

**Pull request merged**